cmake: sysbuild PRE_CMAKE, POST_CMAKE, PRE_DOMAINS, POST_DOMAINS hooks

tejlmand opened the pull request:
Provide a uniform way for sysbuild modules to create hooks that are
called at specific time during sysbuild CMake configure time.
    
A module can create functions following the scheme:
- \<module-name>_pre_cmake
- \<module-name>_post_cmake
- \<module-name>_pre_domains
- \<module-name>_post_domains
    
those functions, if defined, will be called by sysbuild CMake at the
location indicated by the function name.
    
A new global variable `SYSBUILD_CURRENT_MODULE_NAME` is created, which
a module can use to know it's own name when defining those functions
during sysbuild module CMake inclusion.

tejlmand committed:
cmake: provide CMake helper macros for argument validation

This commit provides a number of helper macros for verifying arguments
passing to function:
- zephyr_check_flags_required
- zephyr_check_arguments_required
- zephyr_check_arguments_required_all
- zephyr_check_flags_exclusive
- zephyr_check_arguments_exclusive

which facilitates checking whether a required argument or flag has been
given or if mutual exclusive arguments are given.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: sysbuild PRE_CMAKE, POST_CMAKE, PRE_DOMAINS, POST_DOMAINS hooks

Provide a uniform way for sysbuild modules to create hooks that are
called at specific time during sysbuild CMake configure time.

A module can create functions following the scheme:
- <module-name>_pre_cmake
- <module-name>_post_cmake
- <module-name>_pre_domains
- <module-name>_post_domains

those functions, if defined, will be called by sysbuild CMake at the
location indicated by the function name.

A new global variable `SYSBUILD_CURRENT_MODULE_NAME` is created, which
a module can use to know it's own name when defining those functions
during sysbuild module CMake inclusion.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
DNM as docs are missing

nordicjm reviewed (dismissed):
Minor nit. looks good!

tejlmand committed:
doc: document the sysbuild PRE/POST_CMAKE, PRE/POST_DOMAINS hooks

Document how sysbuild modules can provide functions that will be invoked
during sysbuild CMake configure time.

The following functions has been documented:
- <module-name>_pre_cmake
- <module-name>_post_cmake
- <module-name>_pre_domains
- <module-name>_post_domains

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
